### ISSUE TITLE
mini batch does not work for HDP

### ISSUE THREAD

**Issue by yiruiliu110** [bnpy/bnpy#39]
I passa n_batch =4022 to tge GDPTopic Mdoel as

trained_model, info_dict = bnpy.run(
dataset, 'HDPTopicModel', 'Mult', 'memoVB',
output_path='/tmp/wiki/trymoves-model=hdp_topic+mult-K=5/',
nLap=20, convergeThr=0.01, nBatch=4022,
K=5, initname='randomlikewang',
gamma=50.0, alpha=0.5, lam=0.1,
moves='birth,merge,shuffle',
**dict(list(local_step_kwargs.items()) +
list(merge_kwargs.items()) +
list(birth_kwargs.items())))

the outputs start with:
Dataset Summary:
BagOfWordsData
total size: 1029390 units
batch size: 256 units
num. batches: 4022
Allocation Model: HDP model with K=0 active comps. gamma=50.00. alpha=0.50
Obs. Data Model: Multinomial over finite vocabulary.
Obs. Data Prior: Dirichlet over finite vocabulary

However, the training is extremely slow and takes several hours for an iteration. It seems that the nBatch does not really affect the tranning process.

**Comment by michaelchughes** [comment]
Thanks for reaching out and your interest in our package.

I can assure you that the minibatch learning is working as intended.

I've updated our docs to explain *how* exactly minibatch learning works, and provide some hints and tips.
https://bnpy.readthedocs.io/en/latest/examples/07_wiki/plot-02-demo=scalable_topic_model-model=hdp_topic+mult.html#train-hdp-topic-model

Long story short:
* What is true is that in the same amount of wallclock time, our minibatch algorithms with more batches can often improve the loss (on training set, if not also on validation set) more than fewer batches. See the demo for an example.

* There is a sweet spot in the number of batches, you don't want too small or too large. I don't think doing less than 200 documents per batch makes sense. I would tend toward 1000 or more documents per batch potentially unless you are memory constrained. I think that's a better balance of speed and noise-free updates in the first pass.

* Using more batches will not make the same number of "laps" (aka passes thru the training set, often called "epochs" elsewhere) go faster. This is a myth. This is not true of SGD, and is not true of our incremental/memoized algorithms either. All these methods traverse batches one at a time, and then perform a global update of parameters. More batches means strictly more work per "lap". So make sure you aren't counting by lap but by wallclock time.

**Comment by michaelchughes** [comment]
If you have a very large dataset, I would suggest:

* Adjusting nCoordAscentItersLP and convThrLP as suggested in the docs linked above. You might benefit from making these less strict (fewer iters, less strict convergence thresholds) so you can process more data.
* Adjusting saveEvery, printEvery, and traceEvery to record stats every 0.05 laps (5% of entire dataset) or something. So you aren't waiting forever between reports about progress. With your volume of data, you could see decent models even before you make one full pass thru the dataset.



Birth moves and merge moves often increase the run time substantially (working on this!). 
If your number one goal is speed, you might disable these until you find the right settings of other parameters to make things fast.
